### ISSUE TITLE
ChucK <==> Gwion Interoperability

### ISSUE THREAD

**Issue by scalarwaves** [tonal-glyph/gwiuck#1]
Per [this issue](https://github.com/fennecdjay/Gwion/issues/112) we have a dedicated place to discuss this possibility.

**Comment by fennecdjay** [comment]
Very nice.
Each language already has ways to generate Ast (per se).
We need a converter and a linting module for each.
As a bonus we get a linter for each :smile:

We could somehow achieve it for C/C++ code by other means.
I have the start of a [Swig](www.swig.org) module, gotta clean it and put it on the web.

**Comment by scalarwaves** [comment]
I noticed you reused a lot of type names from Chuck, that should make things easier. Swig seems like a good start

**Comment by fennecdjay** [comment]
Hi, it's been a while!
I may have some time in october to work on this, are you still interested?
If so, would outputting the new code to stdout be the way? it seems quite unixy.

**Comment by scalarwaves** [comment]
Yes, definitely. Since I tend to use Linux as my primary OS, outputting the generated code to STDOUT with a command line flag or something makes sense to me. That way it could be piped into ChucK. There's no 1 - 1 conversion for chugins/ugens, by the way I've been wondering, how are you handling @paulbatchelor removing the soundpipe repo?

**Comment by fennecdjay** [comment]
Can you pipe into chuck? Just tried in my up-to-date git copy and it doesn't seem to work.
I recently added this capability to Gwion, you can now do:
``` sh
echo '<<< "Hello, command line" >>>' | gwion
```
and get the expected result.
However, it's not to hard a write a file and feed chuck with it.
So where good for stdout, nice!
You're we won't be able to have a 1 -1 coversion for various Objects, including, most importantly sound generators. But If STK is ported to Gwion, we could add a few rules for them to match.
Similarly, Chugins could be somehow converted to gwion UsrUGen, just cat the `tick` function and convert it to a `@tick` operator.

I don't really have issue with the removing of soundpipe, there are enough backups on github already. At some point I my ask @PaulBatchelor what he thinks (he joined the github team so I can reach him easily).

**Comment by PaulBatchelor** [comment]
I recommend using a local copy or fork of soundpipe. Any repo that has a
commit within the last 2-3 years should be more than fine. AudioKit's copy
should be reasonably up to date.

**Comment by fennecdjay** [comment]
Nice! I didn't notice the AudioKit copy, got to check it.

**Comment by fennecdjay** [comment]
A have another question: should we provide two binaries or one?
What do you think is better:
  * `gwiuck`·
  * `chuck2gwion` and `gwion2chuck` (alternatively `ck2gw` and 'gw2ck`).


I toyed with chuck grammar, and it appears that if we provide a few functions, we can get a parser without importing everything.
Gwion already has `libgwion-ast` so we should be fine with that.

**Comment by scalarwaves** [comment]
My personal opinion is two binaries. I always liked the x2x form of transforming to and from.

**Comment by fennecdjay** [comment]
Good.
It'll be far easier. I think the way to go is generate an ast, then iterate the ast and output syntax in the other language. Nice and (hopefully easy).

**Comment by fennecdjay** [comment]
Is it OK if I start pushing some code on this repo? (In a PR form, I mean :smile:)
I think I can have something that lends a chuck ast soon.
Gwion'ast problem is already solved by gwion-ast.

**Comment by scalarwaves** [comment]
You have Write access, please by all means. If it's in PR I'll try to keep up but I'm working 6 days a week at the moment.

**Comment by scalarwaves** [comment]
What I mean is, if you like you can push directly to the master branch, or you can submit a pull request, whatever is easiest for you.

**Comment by fennecdjay** [comment]
I got it, thanks.
I'm on my way for `ck2gw` (looking good so far), but I'm taking some time right now to port STK to Gwion.
